### ISSUE TITLE
[luci] Implementation of shape signature inference rule

### ISSUE THREAD

**Issue by llFreetimell** [Samsung/ONE#5037]
# DO NOT WORK THIS FOR A WHILE. IMPORTANT CHANGE WILL BE APPLIED

### Background

This issue is originated from #4372.

### Goal

- Implementation of `ShapeSiganture` inference rule for each `CircleNode`
  - Most of algorithms should be able to be re-used at shape inference algorithm.
- Add testcases for each `ShapeSignature` inference rule
  - You can add already existing code. For example, if there was `Add_000`, you can add `Add_dynamic_000` with `shape_signature`

### How to

1. Select an operation/operations to implement and write your name in the list of table at **assignee**.
    - You can select multiple operations which logic is similar so that PR can be reviewed at once. (ex: ReLU and ReLU6)
    - **Only reserve as much as you can finish in one week. DO NOT BE GREEDY.**
1. Implement each operations as following rules
    - If (operation is virtual)
      - Implement inference rule
    - Else If (operation is circle only)
      - Implement inference rule
      - (If needed, do this first) Add the recipe to `common-artifacts/exclude.lst` if needed.
      - Add testcases to `res/CircleRecipes`
    - Else
      - Implement inference rule
      - (If needed, do this first) Add the recipe to `common-artifacts/exclude.lst` if needed.
      - Add testcases to `res/TensorFlowLiteRecipes`
1. Create PR(s). Then write the PR number in the table at
    - **PR(rule)**, for inference rule implementation
    - **PR(exclude)**, for adding recipe to `common-artifacts/exclude.lst`. If not needed, put :no_entry_sign:.
    - **PR(res)**, for adding recipe to `res/`. **If some recipe is needed to be added to `exclude.lst`, that should be done first**
1. After PR is merged, mark :heavy_check_mark: in the table at **Status**

__NOTE__ You can define a new helper function which can be used in another operation's inference rule.

### Examples (Guide)

- Normal operation : #5073(rule) / #5072(exclude) --> #5079(res)
- Virtual operation : #5039(rule) / exclude, res : Not Needed

**Comment by llFreetimell** [comment]
## Operation Table

| Operation                  | Assignee               | PRs    | Status |
|----------------------------|------------------------|--------|--------|
| Abs |      |      |      |
| Add |      |      |      |
| AddN |      |      |      |
| ArgMax |      |      |      |
| ArgMin |      |      |      |
| AveragePool2D |      |      |      |
| BatchMatMul |      |      |      |
| BatchToSpaceND |      |      |      |
| Cast |      |      |      |
| Ceil |      |      |      |
| Concatenation |      |      |      |
| Const |      |      |      |
| Conv2D |      |      |      |
| Cos |      |      |      |
| Custom |      |      |      |
| DepthToSpace |      |      |      |
| DepthwiseConv2D |      |      |      |
| Dequantize |      |      |      |
| Div |      |      |      |
| Elu |      |      |      |
| Equal |      |      |      |
| Exp |      |      |      |
| ExpandDims |      |      |      |
| Fill |      |      |      |
| Floor |      |      |      |
| FloorDiv |      |      |      |
| FloorMod |      |      |      |
| FullyConnected |      |      |      |
| Gather |      |      |      |
| GatherNd |      |      |      |
| Greater |      |      |      |
| GreaterEqual |      |      |      |
| If |      |      |      |
| L2Normalize |      |      |      |
| L2Pool2D |      |      |      |
| LeakyRelu |      |      |      |
| Less |      |      |      |
| LessEqual |      |      |      |
| LocalResponseNormalization |      |      |      |
| Log |      |      |      |
| LogicalAnd |      |      |      |
| LogicalNot |      |      |      |
| LogicalOr |      |      |      |
| Logistic |      |      |      |
| LogSoftmax |      |      |      |
| MatrixDiag |      |      |      |
| MatrixSetDiag |      |      |      |
| Maximum |      |      |      |
| MaxPool2D |      |      |      |
| Mean |      |      |      |
| Minimum |      |      |      |
| MirrorPad |      |      |      |
| Neg |      |      |      |
| NonMaxSuppressionV4 |      |      |      |
| NonMaxSuppressionV5 |      |      |      |
| NotEqual |      |      |      |
| Pack |      |      |      |
| Pad |      |      |      |
| PadV2 |      |      |      |
| Pow |      |      |      |
| PRelu |      |      |      |
| Range |      |      |      |
| Rank |      |      |      |
| Mul |      |      |      |
| OneHot |      |      |      |
| ReduceAny |      |      |      |
| ReduceMax |      |      |      |
| ReduceMin |      |      |      |
| ReduceProd |      |      |      |
| Relu |      |      |      |
| Relu6 |      |      |      |
| ReluN1To1 |      |      |      |
| Reshape |      |      |      |
| ResizeBilinear |      |      |      |
| ResizeNearestNeighbor |      |      |      |
| ReverseSequence |      |      |      |
| ReverseV2 |      |      |      |
| Round |      |      |      |
| Rsqrt |      |      |      |
| ScatterNd |      |      |      |
| SegmentSum |      |      |      |
| Select |      |      |      |
| SelectV2 |      |      |      |
| Shape |      |      |      |
| Sin |      |      |      |
| Slice |      |      |      |
| Softmax |      |      |      |
| SpaceToBatchND |      |      |      |
| SpaceToDepth |      |      |      |
| SparseToDense |      |      |      |
| Split |      |      |      |
| SplitV |      |      |      |
| Sqrt |      |      |      |
| Square |      |      |      |
| SquaredDifference |      |      |      |
| Squeeze |      |      |      |
| StridedSlice |      |      |      |
| Sub |      |      |      |
| Sum |      |      |      |
| Tanh |      |      |      |
| Tile |      |      |      |
| TopKV2 |      |      |      |
| Transpose |      |      |      |
| TransposeConv |      |      |      |
| UnidirectionalSequenceLSTM |      |      |      |
| Unique |      |      |      |
| Unpack |      |      |      |
| Where |      |      |      |
| While |      |      |      |
| ZerosLike |      |      |      |
| |      |      |      |
| **CircleOnly** | ---- | ---- | ---- |
| BCQFullyConnected |      |      |      |
| BCQGather |      |      |      |
| InstanceNorm |      |      |      |
| |      |      |      |
| **Virtual** | ---- | ---- | ---- |
| Input |      |      |      |
| Output |      |      |      |
| OutputDummy |      |      |      |
| OutputExclude |      |      |      |
| CustomOut |      |      |      |
| IfOut |      |      |      |
| NonMaxSuppressionV4Out |      |      |      |
| NonMaxSuppressionV5Out |      |      |      |
| SplitOut |      |      |      |
| SplitVOut |      |      |      |
| TopKV2Out |      |      |      |
| UniqueOut |      |      |      |
| UnpackOut |      |      |      |
| WhileOut |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (D-F)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| DepthToSpace |      |      |      |      |      |
| DepthwiseConv2D |      |      |      |      |      |
| Dequantize |      |      |      |      |      |
| Div |      |      |      |      |      |
| Elu |      |      |      |      |      |
| Equal |      |      |      |      |      |
| Exp |      |      |      |      |      |
| ExpandDims |      |      |      |      |      |
| Fill |      |      |      |      |      |
| Floor |      |      |      |      |      |
| FloorDiv |      |      |      |      |      |
| FloorMod |      |      |      |      |      |
| FullyConnected |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (G-L)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Greater |      |      |      |      |      |
| GreaterEqual |      |      |      |      |      |
| If |      |      |      |      |      |
| L2Normalize |      |      |      |      |      |
| L2Pool2D |      |      |      |      |      |
| LeakyRelu |      |      |      |      |      |
| Less |      |      |      |      |      |
| LessEqual |      |      |      |      |      |
| LocalResponseNormalization |      |      |      |      |      |
| Log |      |      |      |      |      |
| LogicalAnd |      |      |      |      |      |
| LogicalNot |      |      |      |      |      |
| LogicalOr |      |      |      |      |      |
| Logistic |      |      |      |      |      |
| LogSoftmax |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (M-P)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| MatrixDiag |      |      |      |      |      |
| MatrixSetDiag |      |      |      |      |      |
| Maximum |      |      |      |      |      |
| MaxPool2D |      |      |      |      |      |
| Mean |      |      |      |      |      |
| Minimum |      |      |      |      |      |
| MirrorPad |      |      |      |      |      |
| Mul |      |      |      |      |      |
| Neg |      |      |      |      |      |
| NonMaxSuppressionV4 |      |      |      |      |      |
| NonMaxSuppressionV5 |      |      |      |      |      |
| NotEqual |      |      |      |      |      |
| OneHot |      |      |      |      |      |
| Pack |      |      |      |      |      |
| Pad |      |      |      |      |      |
| PadV2 |      |      |      |      |      |
| Pow |      |      |      |      |      |
| PRelu |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (R)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Range |      |      |      |      |      |
| Rank |      |      |      |      |      |
| ReduceAny |      |      |      |      |      |
| ReduceMax |      |      |      |      |      |
| ReduceMin |      |      |      |      |      |
| ReduceProd |      |      |      |      |      |
| Relu |      |      |      |      |      |
| Relu6 |      |      |      |      |      |
| ReluN1To1 |      |      |      |      |      |
| Reshape |      |      |      |      |      |
| ResizeBilinear |      |      |      |      |      |
| ResizeNearestNeighbor |      |      |      |      |      |
| ReverseSequence |      |      |      |      |      |
| ReverseV2 |      |      |      |      |      |
| Round |      |      |      |      |      |
| Rsqrt |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (S)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| ScatterNd |      |      |      |      |      |
| SegmentSum |      |      |      |      |      |
| Select |      |      |      |      |      |
| SelectV2 |      |      |      |      |      |
| Shape |      |      |      |      |      |
| Sin |      |      |      |      |      |
| Slice |      |      |      |      |      |
| Softmax |      |      |      |      |      |
| SpaceToBatchND |      |      |      |      |      |
| SpaceToDepth |      |      |      |      |      |
| SparseToDense |      |      |      |      |      |
| Split |      |      |      |      |      |
| SplitV |      |      |      |      |      |
| Sqrt |      |      |      |      |      |
| Square |      |      |      |      |      |
| SquaredDifference |      |      |      |      |      |
| Squeeze |      |      |      |      |      |
| StridedSlice |      |      |      |      |      |
| Sub |      |      |      |      |      |
| Sum |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (T-Z)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Tanh |      |      |      |      |      |
| Tile |      |      |      |      |      |
| TopKV2 |      |      |      |      |      |
| Transpose |      |      |      |      |      |
| TransposeConv |      |      |      |      |      |
| UnidirectionalSequenceLSTM |      |      |      |      |      |
| Unique |      |      |      |      |      |
| Unpack |      |      |      |      |      |
| Where |      |      |      |      |      |
| While |      |      |      |      |      |
| ZerosLike |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (Circle Only Ops)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| BCQFullyConnected |      |      |      |      |      |
| BCQGather |      |      |      |      |      |
| InstanceNorm |      |      |      |      |      |

**Comment by llFreetimell** [comment]
## Operation Table (Virtual Ops)

:no_entry_sign: : Not needed
:heavy_check_mark: : Done

| Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
|----------------------------|------------------------|----------|-------------|------|--------|
| Input |  @llFreetimell     | #5039  | :no_entry_sign: | :no_entry_sign: |      |
| Output |   @llFreetimell    | #5039 | :no_entry_sign: | :no_entry_sign: |      |
| OutputDummy |      |      | :no_entry_sign: | :no_entry_sign: |      |
| OutputExclude |      |      | :no_entry_sign: | :no_entry_sign: |      |
| CustomOut |      |      | :no_entry_sign: | :no_entry_sign: |      |
| IfOut |      |      | :no_entry_sign: | :no_entry_sign: |      |
| NonMaxSuppressionV4Out |      |      | :no_entry_sign: | :no_entry_sign: |      |
| NonMaxSuppressionV5Out |      |      | :no_entry_sign: | :no_entry_sign: |      |
| SplitOut |      |      | :no_entry_sign: | :no_entry_sign: |      |
| SplitVOut |      |      | :no_entry_sign: | :no_entry_sign: |      |
| TopKV2Out |      |      | :no_entry_sign: | :no_entry_sign: |      |
| UniqueOut |      |      | :no_entry_sign: | :no_entry_sign: |      |
| UnpackOut |      |      | :no_entry_sign: | :no_entry_sign: |      |
| WhileOut |      |      | :no_entry_sign: | :no_entry_sign: |      |

**Comment by brainer3220** [comment]
> ## Operation Table (A-C)
> 
> :no_entry_sign: : Not needed
> :heavy_check_mark: : Done
> 
> | Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
> |----------------------------|------------------------|----------|-------------|------|--------|
> | Abs |      |      |      |      |      |
> | Add |      |      |      |      |      |
> | AddN |      |      |      |      |      |
> | ArgMax |      |      |      |      |      |
> | ArgMin |      |      |      |      |      |
> | AveragePool2D |      |      |      |      |      |
> | BatchMatMul |      |      |      |      |      |
> | BatchToSpaceND |      |      |      |      |      |
> | Cast |      |      |      |      |      |
> | Ceil |      |      |      |      |      |
> | Concatenation |      |      |      |      |      |
> | Const |      |      |      |      |      |
> | Conv2D | @brainer3220     |      |      |      |      |
> | Cos |      |      |      |      |      |
> | Custom |      |      |      |      |      |

**Comment by brainer3220** [comment]
> ## Operation Table (A-C)
> 
> :no_entry_sign: : Not needed
> :heavy_check_mark: : Done
> 
> | Operation                  | Assignee               | PR(Rule) | PR(exclude) | PR(res) | Status |
> |----------------------------|------------------------|----------|-------------|------|--------|
> | Abs |      |      |      |      |      |
> | Add |      |      |      |      |      |
> | AddN |      |      |      |      |      |
> | ArgMax |      |      |      |      |      |
> | ArgMin |      |      |      |      |      |
> | AveragePool2D |      |      |      |      |      |
> | BatchMatMul |      |      |      |      |      |
> | BatchToSpaceND |      |      |      |      |      |
> | Cast |      |      |      |      |      |
> | Ceil |      |      |      |      |      |
> | Concatenation |      |      |      |      |      |
> | Const |      |      |      |      |      |
> | Conv2D |  @brainer3220    |      |      |      |      |
> | Cos |      |      |      |      |      |
> | Custom |      |      |      |      |      |

**Comment by llFreetimell** [comment]
@brainer3220 You can edit original table :)

**Comment by llFreetimell** [comment]
This issue is outdated and new issue will be created.
Sorry for confusion.

Close this issue!
